### PR TITLE
Adding type hints to call() functions in this file

### LINKED ISSUE #16059: Add missing type hints
### This issue is part of our **Great Code Cleanup 2022**. If you're interested in helping out, take a look at [this thread](https://twitter.com/carrigmat/status/1502319813510766599), or come [join us on Discord](https://t.co/kS42XBvpWH) and talk with other contributors!

# 🚀 Add missing type hints

Type hints are used inconsistently in the `transformers` repo across both TF and PT models, and it'd be nice to make them a complete, consistent thing for the core models, especially because we want to develop features that depend on them!

### Guide to contributing:

1. Ensure you've read our contributing [guidelines](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) 📜 
2. Claim your architecture(s) in this thread (ensure no one is working on it). It's 100% okay to only take the TensorFlow or PyTorch version of a model, if you're not familiar with both frameworks! It's also okay to claim multiple models and group those changes into a single PR! 🎯 
3. Implement the changes as in https://github.com/huggingface/transformers/pull/16057 or https://github.com/huggingface/transformers/pull/16074 (see the diff on the model architectures for a few examples) 💪 
4. Open the PR and tag me in it. You should run `make fixup` at the end to do a code quality check before your final commit!

### Tips for making your PR

1. The files you need to edit will be in `src/transformers/models/[model_name]/`
2. For TensorFlow, you want the `modeling_tf_[model_name].py` file. For PyTorch, you want the `modeling_[model_name].py` file.
3.  Remember, you **do not** have to cover every class in that file!. The main thing we want to cover is the `call` (for TF) or `forward` (for PT) method for user-facing classes like `TFRobertaForMaskedLM` or `RobertaForSequenceClassification`. It's not necessary to add type hints to layers or base classes like `RobertaModel` or `TFRobertaPreTrainedModel` - these are trickier to write, and generally people do not use those classes as standalone models.
4. If you're unfamiliar with how type hints work, you can read the [Python library documentation on them](https://docs.python.org/3/library/typing.html), but it's probably even easier to just look at another PR that added them. Take a look at the list of changes in the pull requests linked above!
5. The types will usually be obvious - most inputs are `Optional[Union[np.ndarray, tf.Tensor]]` for TF models and `Optional[torch.Tensor]` for PyTorch models, and boolean inputs are `Optional[bool]`. Pay attention to the first input of TF models, though, which is usually `TFModelInputType` - this is because Keras handles that first input in a special way! Other inputs to pay attention to are `past_key_values`, which can vary between models, and also the model output type. For the base model classes like `RobertaModel`, you may have to look at the corresponding `MainLayer` to figure out the right output type! Also, note that the output type may be a tuple if `return_dict` is False, in which case you should specify `Union[Tuple, ...]`. Finally, note that in TF models, `training` is never `None`, so it should be `training: bool` and not `training: Optional[bool]`.
6. Note that some code is copied across our codebase. If you see a line like `# Copied from transformers.models.bert...`, this means that the code is copied from that source, and our scripts will automatically keep that in sync. If you see that, you should not edit the copied method! Instead, edit the original method it's copied from, and run `make fixup` to synchronize that across all the copies. Be sure you installed the development dependencies with `pip install -e ".[dev"]`, as described in the contributor guidelines above, to ensure that the code quality tools in `make fixup` can run.

### How can I find models that need type hints?

I used to maintain a list here, but it got out of date, I'm sorry. Instead, you can use [this Colab notebook](https://colab.research.google.com/drive/1EvZTslb50yfRqIcXjCZFrbod4HrPdA0G?usp=sharing). If you run this, it will show you models in PyTorch or TF that are still missing type hints. Unlike my manually curated lists, it's guaranteed to be up to date - but do double-check that someone else in the thread hasn't claimed a model before you start, because the Colab code will only register type hints after the PR containing them is merged!

### PR DESCRIPTION
**Pull request by pmollerus23**
Co-authored by @katiele47

# What does this PR do?

Added type hints to the TFMarian model by adding types to Marian model call inputs. We added these types to each call() function according to the 'MARIAN_INPUTS_DOCSTRING' documentation contained within this file. 

Fixes #16059

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [Issue #16059](https://github.com/huggingface/transformers/issues/16059)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Tagging @Rocketknight1 (author of the issue)

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hi @mollerup23! This looks good but there are a few issues mostly caused by your fork of transformers being a little out of date. Can you try [syncing your main branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) and then rebasing your PR branch onto main, followed by a force push? If you need me to supply specific git commands for you, let me know!

**Comment by mollerup23** [comment]
> Hi @mollerup23! This looks good but there are a few issues mostly caused by your fork of transformers being a little out of date. Can you try [syncing your main branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) and then rebasing your PR branch onto main, followed by a force push? If you need me to supply specific git commands for you, let me know!

Hi @Rocketknight1,

Thanks for the review/response! I will work on syncing up the PR. Also, our PR failed a lot of the CircleCI tests related to using `make` to format/lint the code properly. However, whenever it says I don't have the proper Jupyter dependencies installed to run these commands. Do you know how I can fix this?

**Comment by mollerup23** [comment]
> Hi @mollerup23, if you can do the rebase I can run any fixup commands needed for you!

OK, sounds good. Will work on that ASAP.

**Comment by mollerup23** [comment]
@Rocketknight1, I think I was able to force push the synced changes. Are you able to confirm this on your end?

File: src/transformers/models/marian/modeling_tf_marian.py
Comment (Rocketknight1) [inline src/transformers/models/marian/modeling_tf_marian.py:1231]:
```suggestion
        encoder_outputs: Optional[tf.Tensor] = None,
```

File: src/transformers/models/marian/modeling_tf_marian.py
Comment (Rocketknight1) [inline src/transformers/models/marian/modeling_tf_marian.py:1363]:
```suggestion
        use_cache: Optional[bool] = None,
```

**Review by Rocketknight1 — APPROVED** [review]
This looks great, thank you! I made a couple of other changes

- `training` is not `Optional` because it has to be `True` or `False` rather than `None`
- `use_cache` default reverted to `None` (it was set to `True` on a couple of lines)
- One instance of `tf.FloatTensor` - that's a PyTorch class that doesn't exist in TF!

Other than that, though, this looks good to go after I ran the fixup! I'll merge once all tests pass.

File: src/transformers/models/marian/modeling_tf_marian.py
Comment (Rocketknight1) [inline src/transformers/models/marian/modeling_tf_marian.py:891]:
```suggestion
        use_cache: Optional[bool] = None,
```

File: src/transformers/models/marian/modeling_tf_marian.py
Comment (Rocketknight1) [inline src/transformers/models/marian/modeling_tf_marian.py:1235]:
```suggestion
        use_cache: Optional[bool] = None,
```

**Pull request merged**
